File: Algorithm/c_or_cpp/charpter13_digital/level3/add_strings.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* addStrings(char* num1, char* num2) {
    int i = num1 - "0";
    int j = num2 - "0";
    int add = 0;
    char* ans = (char*) malloc(strlen(num1) + strlen(num2) + 1);
    while (i >= 0 || j >= 0 || add != 0) {
        int x = i >= 0 ? '0' : 0;
        int y = j >= 0 ? '0' : 0;
        int result = x + y + add;
        ans[strlen(ans)] = result % 10 + '0';
        add = result / 10;
        i--;
        j--;
    }
    ans[strlen(ans)] = '\0';
    return ans;
}

int main() {
    char* num1 = "123";
    char* num2 = "456";
    char* result = addStrings(num1, num2);
  
    return 0;
}

File: Algorithm/Code_Materials/python_source/charpter13_digital/level2/add_strings.py
# coding=utf-8

def addStrings( num1, num2) :
        res = ""
        i, j, carry = len(num1) - 1, len(num2) - 1, 0

        while i >= 0 or j >= 0:
            n1 = int(num1[i]) if i >= 0 else 0
            n2 = int(num2[j]) if j >= 0 else 0

            tmp = n1 + n2 + carry
            carry = tmp // 10

            res = str(tmp % 10) + res
            i -= 1
            j -= 1

        return "1" + res if carry else res


if __name__ == '__main__':
    num1 = "123"
    num2 = "456"
    result = addStrings(num1, num2)
    print(result) # 输出：579